File: typescript/bob/bob.ts
class Bob {

    hey(speech: string): string {
        speech = speech.trim()
        if (this.isEmpty(speech)) {
            return 'Fine. Be that way!'
        }
        if (this.isYellQuestion(speech)) {
            return 'Calm down, I know what I\'m doing!'
        }
        if (this.isYelling(speech)) {
            return 'Whoa, chill out!'
        }
        if (this.isQuestion(speech)) {
            return 'Sure.'
        }
        return 'Whatever.'
    }

    isEmpty(speech: string): boolean {
        return !speech.length
    }

    isQuestion(speech: string): boolean {
        return speech.endsWith("?")
    }

    isYelling(speech: string): boolean {
        return (/([A-Z])/.test(speech) && !/([a-z])/.test(speech))
    }

    isYellQuestion(speech: string): boolean {
        return this.isQuestion(speech) && this.isYelling(speech)
    }

}

export default Bob

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"regexp"
	"strings"
)

func hasAlpha(s string) bool {
	reg := regexp.MustCompile(`[a-zA-Z]+`)
	return reg.MatchString(s)
}

func isYelling(s string) bool {
	if !hasAlpha(s) {
		return false
	}
	upper := strings.ToUpper(s)
	return upper == s
}

func isQuestion(s string) bool {
	if len(s) == 0 {
		return false
	}
	return s[len(s)-1] == '?'
}

func isYellingQustion(s string) bool {
	return isQuestion(s) && isYelling(s) && hasAlpha(s)
}

func isSilence(s string) bool {
	return len(s) == 0
}

// Hey responses to the question with limited answers.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)
	if isSilence(remark) {
		return "Fine. Be that way!"
	}
	if isYellingQustion(remark) {
		return "Calm down, I know what I'm doing!"
	}
	if isQuestion(remark) {
		return "Sure."
	}
	if isYelling(remark) {
		return "Whoa, chill out!"
	}
	return "Whatever."
}
